### PR TITLE
perf: instance types cache

### LINKED ISSUE #226: perf: cache fully initialized instance types
### Tell us about your request

Fully initialized instance types (as returned by instanceTypeProvider.List) need to be cached. 

* This needs to be done with a new cache, which would cache the _fully initialized_ instance type list for a given set of inputs. (A convenient way to think about it is "per NodePool", but in reality, a set of inputs may be shared between different nodepools.)
* The existing lower-level cache (in getInstanceTypes) should stay the same, representing all (input-independent) instance types Karpenter can work with. It is there essentially to reduce skewer / resource API calls. It should also have a version (facilitated by change monitor), that input-specific / fully initialized instance type cache can use.
* All inputs affecting the resulting set of instance types need to be considered when caching. There are fields from NodePool (kubelet config) and AKSNodeClass (ImageFamily, OSDiskSize) that need to be hashed/included in the key. (There may be others, but those stand out.)
* Note that the (version of) unavailable offerings cache is yet another input, and should be included as well. (We need to add versioning support to that cache).

Here as well, AWS provider implementation can serve as a reference. Note the implementation of cache versioning (via SeqNum; facilited by ChangeMonitor where needed), composition of the synthetic key, etc.

To demonstrate improvement, it would be good to capture before and after numbers (e.g. maybe just via GetInstanceTypes method duration metric? pprof CPU and memory profile?) Adding a benchmark test (that would also catch regressions) would be great, but could be done separately.

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

Need to avoid having to compose / reinitialize the full list of instance types (with offerings) every time on List.

### Are you currently working around this issue?

The current implementation works, at the expense of some performance hit (and likely quite a bit of extra garbage & GC)

### Additional Context

_No response_

### Attachments

_No response_

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by tallaxes**
Fixes #226 

**Description**

Introduces improved caching for fully initialized instance types. Please see #226 for details, this PR implements everything on the list there.

**How was this change tested?**

* make presubmit
* perf before/after - using CP method duration metric for GetInstanceTypes

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note
Improved performance and memory footprint via instance type caching
```

### CODE REVIEW

**Review by tallaxes — COMMENTED** [review]
/test

File: pkg/providers/instancetype/instancetypes.go [hunk 1]
```diff
@@ -184,7 +226,13 @@ func (p *Provider) getInstanceTypes(ctx context.Context) (map[string]*skewer.SKU
 		}
 	}
 
-	logging.FromContext(ctx).Debugf("%d SKUs remaining after filtering", len(instanceTypes))
+	if p.cm.HasChanged("instance-types", instanceTypes) {
+		// Only update instanceTypesSeqNun with the instance types have been changed
+		// This is to not create new keys with duplicate instance types option
+		atomic.AddUint64(&p.instanceTypesSeqNum, 1)
+		logging.FromContext(ctx).With(
+			"count", len(instanceTypes)).Debugf("discovered instance types")
```
Comment (Bryce-Soghigian) [inline pkg/providers/instancetype/instancetypes.go:234]: can we use a name other than count? writing queries against the logs would be useful to have a term other than count in the json we can query against. 


InstanceCount maybe?

**Comment by Bryce-Soghigian** [comment]
Would be good to get proper benchmarks of before and after, potentially flagging other areas that could use performance improvements as well.

File: pkg/providers/instancetype/instancetypes.go (hunk 1 above)
Comment (tallaxes) [reply pkg/providers/instancetype/instancetypes.go:234]: Wouldn't it be possible to discern context from other fields? I dislike the idea of having to come up with different names in different contexts ...

Comment (Bryce-Soghigian) [reply pkg/providers/instancetype/instancetypes.go:234]: yeah you could do something like 

FakeEventsTable
| where msg startswith "discovered instance types" 
| project count

but the problem is that count is also a reserved keyword in KQL and other languages. So you would have do to 


FakeEventsTable
| where msg startswith "discovered instance types" 
| project ['count']

for it to actually run iirc

**Comment by tallaxes** [comment]
Before:
image
After:
image

**Comment by tallaxes** [comment]
https://github.com/Azure/karpenter-provider-azure/pull/232 refreshes dev monitoring tools and adds the above Provider Method Duration pane to Karpenter Performance dashboard

**Review by Bryce-Soghigian — APPROVED** [review]

**Review by tallaxes — COMMENTED** [review]
/test

**Review by tallaxes — COMMENTED** [review]
/test

**Pull request merged**